File: core/src/main/kotlin/io/gpxtoign/core/layout/Rect.kt
package io.gpxtoign.core.layout

import kotlin.math.max
import kotlin.math.min

/**
 * An axis-aligned rectangle in the *page frame*: the map rotated so that page-up is the
 * job's chosen bearing. Units are metres on the ground.
 */
data class Rect(val uMin: Double, val vMin: Double, val uMax: Double, val vMax: Double) {
    val width get() = uMax - uMin
    val height get() = vMax - vMin
    val centerU get() = (uMin + uMax) / 2.0
    val centerV get() = (vMin + vMax) / 2.0

    fun contains(u: Double, v: Double) = u >= uMin && u <= uMax && v >= vMin && v <= vMax

    fun expand(by: Double) = Rect(uMin - by, vMin - by, uMax + by, vMax + by)

    fun union(o: Rect) = Rect(
        min(uMin, o.uMin), min(vMin, o.vMin), max(uMax, o.uMax), max(vMax, o.vMax),
    )

    fun intersects(o: Rect) =
        uMin <= o.uMax && o.uMin <= uMax && vMin <= o.vMax && o.vMin <= vMax

    companion object {
        /** A [w] x [h] rectangle centred on the given box. */
        fun centeredOn(
            uMin: Double, vMin: Double, uMax: Double, vMax: Double, w: Double, h: Double,
        ): Rect {
            val cu = (uMin + uMax) / 2.0
            val cv = (vMin + vMax) / 2.0
            return Rect(cu - w / 2.0, cv - h / 2.0, cu + w / 2.0, cv + h / 2.0)
        }
    }
}

File: packages/core/src/layout/rect.ts
/**
 * An axis-aligned rectangle in the *page frame*: the map rotated so that page-up is the
 * job's chosen bearing. Units are metres on the ground.
 */
export interface Rect {
  readonly uMin: number
  readonly vMin: number
  readonly uMax: number
  readonly vMax: number
}

export function rect(uMin: number, vMin: number, uMax: number, vMax: number): Rect {
  return { uMin, vMin, uMax, vMax }
}

export function width(r: Rect): number {
  return r.uMax - r.uMin
}

export function height(r: Rect): number {
  return r.vMax - r.vMin
}

export function centerU(r: Rect): number {
  return (r.uMin + r.uMax) / 2
}

export function centerV(r: Rect): number {
  return (r.vMin + r.vMax) / 2
}

export function contains(r: Rect, u: number, v: number): boolean {
  return u >= r.uMin && u <= r.uMax && v >= r.vMin && v <= r.vMax
}

export function expand(r: Rect, by: number): Rect {
  return rect(r.uMin - by, r.vMin - by, r.uMax + by, r.vMax + by)
}

export function union(a: Rect, b: Rect): Rect {
  return rect(
    Math.min(a.uMin, b.uMin),
    Math.min(a.vMin, b.vMin),
    Math.max(a.uMax, b.uMax),
    Math.max(a.vMax, b.vMax),
  )
}

export function intersects(a: Rect, b: Rect): boolean {
  return a.uMin <= b.uMax && b.uMin <= a.uMax && a.vMin <= b.vMax && b.vMin <= a.vMax
}

/** A `w` x `h` rectangle centred on the given box. */
export function centeredOn(
  uMin: number,
  vMin: number,
  uMax: number,
  vMax: number,
  w: number,
  h: number,
): Rect {
  const cu = (uMin + uMax) / 2
  const cv = (vMin + vMax) / 2
  return rect(cu - w / 2, cv - h / 2, cu + w / 2, cv + h / 2)
}
